### PR TITLE
 Rework toDate and extract parseISO

### PR DESCRIPTION
**Pull request by kossnocorp**
The PR is a result of an attempt to make minimal build size smaller. See https://github.com/date-fns/date-fns/pull/1015 for the background.

- Extract `parseISO` from `toDate`
- Make `toDate` accept only numbers and dates
- Remove global options object
- Fixes DST issue in `eachDayOfInterval` that caused time in the days after DST change to have the shift as well.

The results of the PR:

- `lightFormat`: `1.73 KB` → `887 B`
- `toDate`: `1.27 KB` → `296 B`
- `addDays`: `1.31 KB` → `385 B`

For comparison, here are the v1 stats:

- `format`: `2.77 KB`
- `addDays`: `983 B`

### CODE REVIEW

File: docs/upgradeGuide.md [hunk 1]
```diff
@@ -12,10 +12,22 @@
   import differenceInCalendarISOYears from 'date-fns/differenceInCalendarISOYears'
   ```
 
+- Functions now don't accept strings as arguments. Strings should
+  be parsed using `parseISO` (ISO 8601) or `parse`.
+
+  See [this post](https://blog.date-fns.org/post/TODO) for more details.
```
Comment (leshakoss) [inline docs/upgradeGuide.md]: TODO?

File: docs/strings.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# String Arguments
+
+TODO
```
Comment (leshakoss) [inline docs/strings.md]: todo

File: docs/upgradeGuide.md (hunk 1 above)
Comment (kossnocorp) [reply docs/upgradeGuide.md]: @leshakoss yup, will update once the post is published

File: docs/strings.md (hunk 2 above)
Comment (kossnocorp) [reply docs/strings.md]: @leshakoss damn, I've forgot to remove the doc

File: test/dst/eachDayOfInterval/basic.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  eachDayOfInterval({
+    start: new Date(2020, 2, 26),
+    end: new Date(2020, 2, 28)
+  }).map(d => d.toString()),
+  [
+    'Thu Mar 26 2020 00:00:00 GMT+0200 (Eastern European Standard Time)',
+    'Fri Mar 27 2020 01:00:00 GMT+0300 (Eastern European Summer Time)',
+    'Sat Mar 28 2020 00:00:00 GMT+0300 (Eastern European Summer Time)'
+  ]
+)
```
Comment (kossnocorp) [inline test/dst/eachDayOfInterval/basic.js]: I've messed up with the rebase, so for those who are looking for the actual fix: https://github.com/date-fns/date-fns/pull/1023/files#diff-5a48fe3ae25f8ac62336262cb0acd5e2

**Pull request merged**